ostd/src/mm/page/meta.rs

```diff
@@ -92,196 +70,94 @@ pub(in crate::mm) struct MetaSlot {
-    pub(super) usage: AtomicU8,
+    ///  - if the metadata need special alignment, we can provide
+    ///    at most `PAGE_METADATA_ALIGN` bytes of alignment;
+    ///  - the subsequent fields can utilize the padding of the
+    ///    reference count to save space.
+    storage: UnsafeCell<[u8; PAGE_METADATA_MAX_SIZE]>,
+    /// The reference count of the page.
     pub(super) ref_count: AtomicU32,
+    /// The virtual table that indicates the type of the metadata.
+    pub(super) vtable_ptr: UnsafeCell<PageMetaVtablePtr>,
 }
 
-pub(super) union MetaSlotInner {
-    _frame: ManuallyDrop<FrameMeta>,
-    _pt: ManuallyDrop<PageTablePageMeta>,
-}
+type PageMetaVtablePtr = core::ptr::DynMetadata<dyn PageMeta>;
 
-// Currently the sizes of the `MetaSlotInner` union variants are no larger
-// than 8 bytes and aligned to 8 bytes. So the size of `MetaSlot` is 16 bytes.
-//
-// Note that the size of `MetaSlot` should be a multiple of 8 bytes to prevent
-// cross-page accesses.
-const_assert_eq!(size_of::<MetaSlot>(), 16);
+const_assert_eq!(PAGE_SIZE % META_SLOT_SIZE, 0);
+const_assert_eq!(size_of::<MetaSlot>(), META_SLOT_SIZE);
 
-/// All page metadata types must implemented this sealed trait,
-/// which ensures that each fields of `PageUsage` has one and only
-/// one metadata type corresponding to the usage purpose. Any user
-/// outside this module won't be able to add more metadata types
-/// and break assumptions made by this module.
+/// All page metadata types must implemented this trait.
 ///
 /// If a page type needs specific drop behavior, it should specify
 /// when implementing this trait. When we drop the last handle to
-/// this page, the `on_drop` method will be called.
-pub trait PageMeta: Sync + private::Sealed + Sized {
-    const USAGE: PageUsage;
+/// this page, the `on_drop` method will be called. The `on_drop`
+/// method is called with the physical address of the page.
+///
+/// # Safety
+///
+/// The implemented structure must have a size less than or equal to
+/// [`PAGE_METADATA_MAX_SIZE`] and an alignment less than or equal to
+/// [`PAGE_METADATA_MAX_ALIGN`].
+pub unsafe trait PageMeta: Any + Send + Sync + 'static {
+    fn on_drop(&mut self, _paddr: Paddr) {}
+}
 
-    fn on_drop(page: &mut Page<Self>);
+/// Makes a structure usable as a page metadata.
+///
+/// Directly implementing [`PageMeta`] is not safe since the size and alignment
+/// must be checked. This macro provides a safe way to implement the trait with
+/// compile-time checks.
+#[macro_export]
+macro_rules! impl_page_meta {
+    ($($t:ty),*) => {
+        $(
+            use static_assertions::const_assert;
+            const_assert!(size_of::<$t>() <= $crate::mm::page::meta::PAGE_METADATA_MAX_SIZE);
+            const_assert!(align_of::<$t>() <= $crate::mm::page::meta::PAGE_METADATA_MAX_ALIGN);
+            // SAFETY: The size and alignment of the structure are checked.
+            unsafe impl $crate::mm::page::meta::PageMeta for $t {}
```

Comment:
I think that we should allow the caller of this macro to provide an (optional) implementation of the `on_drop` method. Do you have in mind the first use case where an non-trivial `on_drop` method is implemented?

Reply:
The page table metadata need it and it uses unsafe implementation.

Outside OSTD there will be usages, on dropping page cache pages the content should be written back to the disk. We could put it off for future works.

Reply:
Ok